src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1754,6 +1754,15 @@ private static FrozenArrayBuilder<T> AsFrozen<T>(ArrayBuilder<T> builder)
             public FrozenArrayBuilder(ArrayBuilder<T> arrayBuilder)
             {
                 Debug.Assert(arrayBuilder != null);
+                if (arrayBuilder.Capacity >= ArrayBuilder<T>.PooledArrayLengthLimitExclusive
+                    && arrayBuilder.Count < ArrayBuilder<T>.PooledArrayLengthLimitExclusive
+                    && arrayBuilder.Capacity > arrayBuilder.Count * 2)
```

Comment:
📝 An array builder meeting this condition will satisfy the following:

1. The current backing array is too large to be returned to the pool (it will be GC whenever no longer used)
2. The resized backing array from trimming is small enough to be returned to the pool (i.e. it will not be wasted after this builder is freed)
3. At least half of the storage of the current builder is unused

Reply:
➡️ Added to source